shoop/admin/modules/orders/static_src/create/reducers/lines.js

```diff
@@ -68,24 +106,60 @@ function setLineProperty(state, {payload}) {
                     updates.sku = null;
                 }
                 if (value === "text") {
+                    updates = getDiscountsAndTotal(0, line.baseUnitPrice, 0);
                     updates.unitPrice = 0;
                     updates.quantity = 0;
                 }
+                updates.type = value;
                 break;
             case "quantity":
-                updates.quantity = Math.max(0, parseFloat(value));
+                var quantity = Math.max(0, parseFloat(value));
+                updates  = getDiscountsAndTotal(quantity, line.baseUnitPrice, line.unitPrice);
+                updates.quantity = quantity;
                 break;
             case "unitPrice":
-                updates.unitPrice = Math.max(0, parseFloat(value));
+                var unitPrice = parseFloat(value);
+                updates  = getDiscountsAndTotal(line.quantity, line.baseUnitPrice, unitPrice);
+                updates.unitPrice = unitPrice;
+                break;
+            case "discountPercent":
+                var discountPercent = Math.max(0, parseFloat(value));
+                updates  = getDiscountsAndTotal(
+                    line.quantity, line.baseUnitPrice, (line.baseUnitPrice * (1 - (discountPercent / 100))), true
+                );
+                updates.discountPercent = discountPercent;
+                break;
+            case "discountAmount":
+                var discountAmount = Math.max(0, parseFloat(value));
+                updates  = getDiscountsAndTotal(
+                    line.quantity, line.baseUnitPrice, (line.baseUnitPrice - (discountAmount / line.quantity)), true
+                );
+                updates.discountAmount = discountAmount;
+                break;
+            case "total":
+                var total = parseFloat(value);
+                updates  = getDiscountsAndTotal(line.quantity, line.baseUnitPrice, (total / line.quantity), true);
+                updates.total = total;
                 break;
         }
     }
     return setLineProperties(state, id, updates);
 }
 
+function updateLines(state, {payload}) {
```

Comment:
Actually, this is quite obviously an async action, so this logic should live in an async action creator.

So...

* Get rid of this reducer altogether
* Make the `updateLines` action look like this:

```javascript
const updateLines = () => {
    return (dispatch, getState) => {
        getState().lines.forEach((line) => {
            if (line.product) {
                dispatch(retrieveProductData({id: line.product.id, forLine: line.id, quantity: line.quantity}));
            }
        });
    }
}
```

This way action dispatching happens where it should -- in actions -- and not in the reducer :sunglasses: